File: 20 Valid Parentheses/Python/ValidParenthese.py
from collections import deque

class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        open_parenthesis_deque = deque()
        just_popped_char = ""

        for each_char in s:
            # start a new open
            if (each_char == '(' or each_char == '[' or each_char == '{'):
                open_parenthesis_deque.append(each_char)
            else:
                # nothing on the stack to close, not enough openings
                if (not open_parenthesis_deque):
                    return False

                just_popped_char = open_parenthesis_deque.pop()

                # does the close match the last open?
                if (each_char == ')'):
                    if (just_popped_char != '('):
                        return False
                elif (each_char == ']'):
                    if (just_popped_char != '['):
                        return False
                else:
                    if (just_popped_char != '{'):
                        return False

        # still something on the stack, not enough closings
        if (open_parenthesis_deque):
            return False

        return True


# Various test cases.
# input_value = "()"
# input_value = "()[]{}"
input_value = "(]"
# input_value = "([])"
# input_value = "([)]"
# input_value = "["
# input_value = "]"

# Display the starting value.
print("Input Value: ", input_value)

# Perform the operation on the data.
mySolution = Solution()
output_value = mySolution.isValid(input_value)

# Display the final value.
print("Output Value: ", output_value)

File: 20 Valid Parentheses/C++/ValidParenthese.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> open_parenthesis_stack;
        char popped_char;

        for(char each_char : s)
        {
            // start a new open
            if (each_char == '(' || each_char == '[' || each_char == '{')
            {
                open_parenthesis_stack.push(each_char);
            }
            else
            {
                // nothing on the stack to close, not enough openings
                if (open_parenthesis_stack.size() == 0)
                {
                    return false;
                }

                popped_char = open_parenthesis_stack.top();
                open_parenthesis_stack.pop();

                // does the close match the last open?
                if (each_char == ')')
                {
                    if (popped_char != '(')
                    {
                        return false;
                    }
                }
                else if (each_char == ']')
                {
                    if (popped_char != '[')
                    {
                        return false;
                    }
                }
                else
                {
                    if (popped_char != '{')
                    {
                        return false;
                    }
                }
            }
        }

        // still something on the stack, not enough closings
        if (open_parenthesis_stack.size() != 0)
        {
            return false;
        }

        return true;
    }
};

int main()
{
    // Various test cases.
    // string input_value = "()";
    string input_value = "()[]{}";
    // string input_value = "(]";
    // string input_value = "([])";
    // string input_value = "([)]";
    // string input_value = "[";
    // string input_value = "]";

    // Display the starting value.
    cout << "Input Value: " << input_value << endl;

    // Perform the operation on the data.
    Solution mySolution = Solution();
    bool output_value = mySolution.isValid(input_value);

    // Display the final value.
    cout << "Output Value: " << std::boolalpha << output_value << endl;
}
